packages/apollo-cli/src/commands/codegen/generate.ts

```diff
@@ -0,0 +1,209 @@
+    }),
+    tagName: flags.string({
+      description: "Name of the template literal tag used to identify template literals containing GraphQL queries in Javascript/Typescript code",
+      default: "gql"
+    }),
+  };
+
+  static args = [
+    {
+      name: "output",
+      description: "Path to write the generated code to",
+      required: true
+    }
+  ]
+
+  async run() {
+    const { flags, args } = this.parse(Generate);
+
+    if (!args.output) {
+      this.error("The output path must be specified in the arguments");
+      return;
+    }
+
+    let inferredTarget: TargetType;
+    if (flags.target) {
+      if (["swift", "typescript", "flow", "scala"].includes(flags.target)) {
+        inferredTarget = flags.target as TargetType;
+      } else {
+        this.error(`Unsupported target: ${flags.target}`);
+      }
+    } else {
+      switch(args.output.split('.').reverse()[0]) {
+        case "swift":
+          inferredTarget = "swift";
+          break;
+
+        case "ts" || "tsx":
+          inferredTarget = "typescript";
+          break;
+
+        case "js":
+          inferredTarget = "flow";
+          break;
+
+        case "scala":
+          inferredTarget = "scala";
+          break;
+
+        default:
+          this.error("Could not infer target from output file type, please use --target");
+          return;
+      }
+    }
+
+    const apiKey = process.env.ENGINE_API_KEY || flags.key;
+    const pullFromEngine = !!apiKey;
+
+    const tasks = new Listr([
+      {
+        title: "Scanning for GraphQL queries",
+        task: async (ctx, task) => {
+          const paths = await globby(flags.queries ? [ flags.queries ] : []);
+          task.title = `Scanning for GraphQL queries (${paths.length} found)`;
+          ctx.queryPaths = paths;
+        }
+      },
+      {
+        title: pullFromEngine ? "Loading schema from Apollo Engine" : "Loading GraphQL schema",
+        task: async ctx => {
+          if (pullFromEngine) {
+            const variables = {
+              id: getIdFromKey(apiKey as string),
+              tag: flags.tag || "current",
+            }
+
+            const engineSchema = await toPromise(
+              execute(engineLink, {
+                query: SCHEMA_QUERY,
+                variables,
+                context: {
+                  headers: { ["x-api-key"]: apiKey },
+                  ...(flags.engine && { uri: flags.engine }),
+                },
+              })
+            );
+
+            if (engineSchema.data && engineSchema.data.service.schema) {
+              ctx.schema = buildClientSchema(
+                engineSchema.data.service.schema
+              );
+            } else {
+              this.error("Failed to get schema from Apollo Engine");
+            }
+          } else {
+            const schemaFileContent = await promisify(fs.readFile)(path.resolve(flags.schema as string));
```

Comment:
fyi there is `fs.readFileSync` that you can use here

Reply:
I decided to use the async version so that the spinner keeps spinning while we are reading content (not major but fun 😄)